h2o-algos/src/main/java/hex/tree/gbm/GBM.java

```diff
@@ -916,6 +917,21 @@ private void rollupMinMaxPreds(DTree tree, DTree.Node node, float[] mins, int mi
       return new GBMModel(modelKey, parms, new GBMModel.GBMOutput(GBM.this));
     }
 
+    @Override
+    protected void doInTrainingCheckpoint() {
+      try {
+        String modelFile = _parms._in_training_checkpoints_dir + "/" + _model._key.toString() + "." + _model._output._ntrees;
```

Comment:
summing up our discussion about this, I'd suggest to consolidate the checkpoint logic:
- save in same `_export_checkpoints_dir` for all scenarios.
- only keep the last checkpoint, and save the last transactionally: 
   1. if exists, rename old with `.bak` extension.
   2. save new.
   3. delete old or restore it if #2 failed.
- add some kind of `_completed` property on the model itself so that when we load a checkpointed model, we know if it can be resumed or not (not sure this is needed immediately but could be useful for grids/automl).
- add a `checkpoint_events` (or a better name) param listing when checkpointing should occur. Today, when `export_checkpoints_dir` is set, it does it only on completion, so for example the default value could be:
```java
CheckpointEvent[] checkpoint_events = { onCompletion };
```
and for this PR scenario, we would have sth like:
```java
CheckpointEvent[] checkpoint_events = { onCompletion, onIterations(3) };
```
My idea is that AutoML could easily benefit from checkpointing in the future to save memory together with incremental search (e.g. successive-halving), and I'm suggesting a flexible mechanism that would allow us to easily add more conditions on when this should occur.
- if the checkpointing logic is consolidated, the reloading/recovery logic however is usually specific: from manual, to automatically load all checkpointed models, or for grids for example only load the last one to determine if it was completed or not... but at least we always know that we have the most recent version of the model.
- to test in-training scenario, suggesting to mock GBM driver (ask for 10 trees, checkpoint at every iteration, mock triggers model failure if iteration 6 for example, check that there is a checkpointed model saved with 5 trees).

@michalkurka wdyt? am I overthinking this?

Reply:
Checkpointing is not limited to the local filesystem - it works with any supported storage - I don't think there is rename on S3 for example - https://stackoverflow.com/questions/32501995/boto3-s3-renaming-an-object-using-copy-object

I also think _export_checkpoints_dir serves a different purpose, especially with auto-recovery - I would not mix these 2 concepts.

I would keep the PR conceptually the same and just focus on the implementation.

Reply:
> I also think _export_checkpoints_dir serves a different purpose, especially with auto-recovery - I would not mix these 2 concepts.

I understand that there can be different recovery scenarios, but I don't see why this should mean different saving destination. In all cases, what you want is to have one place with the most recent saved version of your model/grid/..
Or, are you saying that you may want to save completed models for autorecovery on s3 but intermediate ones for Sagemaker on the file system? or the other way round? Would anyone use both?
I find it confusing as it looks like each time the user is specifying a target dir, one needs to know exactly what will go here or there, when it will be saved (this is implicit).

Renaming logic is not needed, I see that we call `exportBinaryModel` with `force=true` to override existing model, so the storage provider should just handle this if it can.

Reply:
> what you want is to have one place with the most recent saved version of your model/grid/..
The issue is that it is not actually a valid model as you would expect it. The metrics do not describe the performance of the model because scoring and checkpointing can be out-of-sync. If you enable both mechanisms you will have 2 different object types in the same directory - which I don't think I would want to. IMHO what you want is to have the finished model in one directory and the still trained in another one because you want to clean up the "in-training" checkpointed one while you want to keep the fully trained one.

Reply:
The other reason is that in my point of view the "in-training" checkpoints just happen to be binary models - current implementation is a starting point and in my view only a shortcut - it might also make sense to store metadata and trees (for example) separately and also store some additionally (normally) transient information - eg. cached p1 predictions.

Either way the checkpointed object is not a valid model that someone can just load and work with.